src/v2/controllers/vdeployment_controller/exec/reconciler.rs

```diff
@@ -0,0 +1,505 @@
+            let spec_filter = |vrs: VReplicaSetView|
+                vrs.metadata.owner_references_contains(vd@.controller_owner_ref())
+                && vrs.metadata.deletion_timestamp.is_none()
+                && vrs.well_formed();
+            let pre_filtered_vrs_list = if spec_filter(vrs@) {
+                filtered_vrs_list@.map_values(|vrs: VReplicaSet| vrs@).drop_last()
+            } else {
+                filtered_vrs_list@.map_values(|vrs: VReplicaSet| vrs@)
+            };
+            assert(pre_filtered_vrs_list == vrs_list@.map_values(|vrs: VReplicaSet| vrs@).take(idx as int).filter(spec_filter));
+            push_filter_and_filter_push(vrs_list@.map_values(|vrs: VReplicaSet| vrs@).take(idx as int), spec_filter, vrs@);
+            assert(vrs_list@.map_values(|vrs: VReplicaSet| vrs@).take(idx as int).push(vrs@)
+                   == vrs_list@.map_values(|vrs: VReplicaSet| vrs@).take(idx + 1 as int));
+            assert(spec_filter(vrs@) ==> filtered_vrs_list@.map_values(|vrs: VReplicaSet| vrs@) == pre_filtered_vrs_list.push(vrs@));
+        }
+
+        idx += 1;
+    }
+    assert(vrs_list@.map_values(|vrs: VReplicaSet| vrs@) == vrs_list@.map_values(|vrs: VReplicaSet| vrs@).take(vrs_list.len() as int));
+    filtered_vrs_list
+}
+
+fn filter_old_and_new_vrs(vrs_list: Vec<VReplicaSet>, vd: &VDeployment) -> (res: (Vec<VReplicaSet>, Vec<VReplicaSet>))
+requires
+    vd@.well_formed(),
+    // vrs.well_formed() is required because we need to update the old vrs -> old_vrs.metadata.well_formed()
+    // and new/old vrs has replicas -> vrs.state_validation()
+    forall |i: int| 0 <= i < vrs_list.len() ==> #[trigger] vrs_list[i]@.well_formed()
+ensures
+    res.0@.map_values(|vrs: VReplicaSet| vrs@) == model_reconciler::filter_old_and_new_vrs(vrs_list@.map_values(|vrs: VReplicaSet| vrs@), vd@).0,
+    res.1@.map_values(|vrs: VReplicaSet| vrs@) == model_reconciler::filter_old_and_new_vrs(vrs_list@.map_values(|vrs: VReplicaSet| vrs@), vd@).1,
+    forall |i: int| 0 <= i < res.0.len() ==> res.0[i]@.well_formed(),
+    forall |i: int| 0 <= i < res.1.len() ==> res.1[i]@.well_formed(),
+{
+    let mut new_vrs_list = Vec::new();
+    let mut old_vrs_list = Vec::new();
+    let mut idx = 0;
+    let pod_template_hash = vd.metadata().resource_version().unwrap();
+
+    proof {
+        assert(
+            new_vrs_list@.map_values(|vrs: VReplicaSet| vrs@) ==
+            model_reconciler::filter_old_and_new_vrs(vrs_list@.map_values(|vrs: VReplicaSet| vrs@).take(0), vd@).0
+        );
+        assert(
+            old_vrs_list@.map_values(|vrs: VReplicaSet| vrs@) ==
+            model_reconciler::filter_old_and_new_vrs(vrs_list@.map_values(|vrs: VReplicaSet| vrs@).take(0), vd@).1
+        );
+        assert(forall |i: int| 0 <= i < new_vrs_list.len() ==> new_vrs_list[i]@.well_formed());
+        assert(forall |i: int| 0 <= i < old_vrs_list.len() ==> old_vrs_list[i]@.well_formed());
+    }
+
+    while idx < vrs_list.len()
```

Comment:
Good job figuring out how to put both new and old lists into a single loop invariant!